### ISSUE TITLE
Implement multi-mixotroph generalization of the NPBD model

### ISSUE THREAD

**Issue by glwagner** [CliMA/ClimaOceanBiogeochemistry.jl#6]
It would seem that the NPBD model might be our first "useful" model? So we should implement that in the source code.

I know @emilyzakem also has aspiration for an "Multi-species NPBD", where I think either P or B can profilerate into NP or NB "species". This model would then have NP + NB + 2 tracers.

Perhaps `MultiSpeciesNutrientsPlanktonBacteriaDetritus` is a good name for that model?

**Comment by emilyzakem** [comment]
Can I first show you a similar version of this model that does have
multiple types to see what you think? It has nn and nd as well as np and
nb. Hopefully that alleviates the "N" name problem too. I'd like to further
think about merging p and b, and maybe even z, since there's lots of
mixotrophs. The biogeochemical roles of populations would then be defined
using matrices of substrate (including light, or other populations, in the
case of grazing) inputs and outputs. I only worry this would get opaque
quickly. Maybe next week I walk through my current Julia model with you?

**Comment by glwagner** [comment]
Yes sure, happy to take a look next week! You can also paste or link to code in this issue to further the discussion.

In terms of implementing this within the current Oceananigans interface, the main challenge is software. Right now we require a _different_ function for each tracer. So here we will have, for example, `model.tracers.mixotroph_1, model.tracers.mixotroph_2`, etc, and we require a different "dynamics function" for each one that dispatches on `::Val{:mixotroph_1}`, `::Val{mixotroph_2}`, etc. 

This challenge is easily overcome however, since we can use metaprogramming to automatically generate such functions. Again more of a julia 102 than julia 101 concept, but nonetheless straightforward in principle.

Once that minor software hurdle is overcome, its trivial to use matrices, tuples, etc to hold any number of parameters that generate different behavior for each mixotroph species within some homogeneous mathematical framework (ie all mixotrophs have their own maximum growth rate, mortality rate, etc).

**Comment by glwagner** [comment]
I changed the title of this issue to reflect the discussion!
